### PR TITLE
[PFS-251] ReadFilesetCDR Performance Improvements

### PR DESCRIPTION
**Pull request by Zhang-Muyang**
SignedURL caching (don’t create new signed urls if we don’t need to within a single RPC)
Parallel creation of signed URLs

### CODE REVIEW

File: src/internal/storage/server.go [hunk 1]
```diff
@@ -220,23 +225,48 @@ func (w *writer) Write(data []byte) (int, error) {
...
+	if err != nil {
+		panic(err)
+	}
+	taskChain := taskchain.New(ctx, semaphore.NewWeighted(int64(taskParallelism)))
+	if err := s.readFileset(ctx, request, func(f fileset.File) error {
 		var refs []*cdr.Ref
-		for _, dr := range f.Index().File.DataRefs {
-			ref, err := s.Chunks.CDRFromDataRef(ctx, dr)
-			if err != nil {
+		refNum := len(f.Index().File.DataRefs)
```
Comment (FahadBSyed) [inline src/internal/storage/server.go:235]: [nit]: refNum implies that each reference has a number (i.e. an id), whereas `numRefs` implies a count of refs.
I'd suggest renaming this to one of `numRefs`, `countRefs`, `refCount`.

File: src/internal/storage/chunk/cdr.go [hunk 2]
```diff
@@ -2,24 +2,31 @@ package chunk
...
 	defer client.Close()
-	path, err := client.GetPath(ctx, ID(dataRef.Ref.Id))
-	if err != nil {
-		return nil, err
-	}
-	url, err := s.bucket.SignedURL(ctx, path, &blob.SignedURLOptions{Expiry: 24 * time.Hour})
-	if err != nil {
-		return nil, errors.EnsureStack(err)
+	key := string(dataRef.Ref.Id)
+	url, ok := cache.Get(key)
```
Comment (FahadBSyed) [inline src/internal/storage/chunk/cdr.go:18]: [nit] `ok` is the correct name for this, but sometimes I like to give `ok` a more intentional name so I don't have to think about what its used for as I'm reading. For example `isCached`

File: src/internal/storage/server_test.go [hunk 3]
```diff
@@ -255,6 +255,18 @@ func createFileset(ctx context.Context, c storage.FilesetClient, num, size int)
 				return "", nil, err
 			}
 		}
+		if size == 0 {
```
Comment (brycemcanally) [inline src/internal/storage/server_test.go:258]: A cleaner way to do this imo is to always send an initial message with zero bytes before the chunking above.

File: src/internal/storage/server.go [hunk 4]
```diff
@@ -220,23 +225,48 @@ func (w *writer) Write(data []byte) (int, error) {
 
 func (s *Server) ReadFilesetCDR(request *storage.ReadFilesetRequest, server storage.Fileset_ReadFilesetCDRServer) error {
 	ctx := server.Context()
-	// TODO: Use task chain here.
-	return s.readFileset(ctx, request, func(f fileset.File) error {
+	cache, err := lru.New[string, string](128)
+	if err != nil {
+		panic(err)
```
Comment (brycemcanally) [inline src/internal/storage/server.go:230]: lets just return the error here.

File: src/internal/storage/server.go [hunk 5]
```diff
@@ -220,23 +225,48 @@ func (w *writer) Write(data []byte) (int, error) {
 
 func (s *Server) ReadFilesetCDR(request *storage.ReadFilesetRequest, server storage.Fileset_ReadFilesetCDRServer) error {
 	ctx := server.Context()
-	// TODO: Use task chain here.
-	return s.readFileset(ctx, request, func(f fileset.File) error {
+	cache, err := lru.New[string, string](128)
```
Comment (brycemcanally) [inline src/internal/storage/server.go:228]: make the cache size a constant

File: src/internal/storage/server.go [hunk 6]
```diff
@@ -220,23 +225,48 @@ func (w *writer) Write(data []byte) (int, error) {
...
-			refs = append(refs, ref)
 		}
-		// TODO: Move to cdr package?
-		ref := chunk.CreateConcatRef(refs)
-		return server.Send(&storage.ReadFilesetCDRResponse{
-			Path: f.Index().Path,
-			Ref:  ref,
-		})
-	})
+		if numRefs == 0 {
```
Comment (brycemcanally) [inline src/internal/storage/server.go:259]: This should be the first check in the callback. Also, I don't think there is really much value in having the `numRefs` variable, just getting the length in the two places it is referenced seems simpler.

**Review by brycemcanally — APPROVED** [review]
lgtm, just one minor nitpick left.

**Pull request merged**